segmentstore/server/src/main/java/io/pravega/segmentstore/server/containers/ContainerEventProcessorImpl.java

```diff
@@ -123,12 +130,32 @@
      * {@link ContainerEventProcessor.EventProcessor} based on its name.
      */
     @VisibleForTesting
-    static Function<String, CompletableFuture<DirectSegmentAccess>> getOrCreateInternalSegment(SegmentContainer container,
-                                                                                               MetadataStore metadataStore,
-                                                                                               Duration timeout) {
-        return s -> metadataStore.registerPinnedSegment(getEventProcessorSegmentName(container.getId(), s),
-                SYSTEM_CRITICAL_SEGMENT, null, timeout) // Segment should be pinned.
-                .thenCompose(l -> container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout));
+    Function<String, CompletableFuture<DirectSegmentAccess>> getOrCreateInternalSegment(SegmentContainer container,
+                                                                                        MetadataStore metadataStore,
+                                                                                        Duration timeout) {
+        return s -> tryCreateAndPinInternalSegment(container, metadataStore, s, timeout)
+                .thenCompose(v -> container.forSegment(getEventProcessorSegmentName(container.getId(), s), timeout));
+    }
+
+    private CompletableFuture<Void> tryCreateAndPinInternalSegment(SegmentContainer container, MetadataStore metadataStore,
+                                                                   String eventProcessorName, Duration timeout) {
+        CompletableFuture<EventProcessor> eventProcessorFuture = this.eventProcessorMap.get(eventProcessorName);
+        if (eventProcessorFuture == null || !eventProcessorFuture.isDone()) {
```

Comment:
Now the thread safety is gone after you remove synchronized. check and put are two separate operations and not atomic anymore

Look into [ConcurrentHashMap::computeIfAbsent](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#computeIfAbsent-K-java.util.function.Function-) on how to do atomic check and add.

Reply:
`computeIfAbsent` does not work here, because we also want to check if the future is present but not complete. Perhaps the [ConcurrentHashMap::compute](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/ConcurrentHashMap.html#compute-K-java.util.function.BiFunction-) could work, but I'm not convinced that we need atomicity here by embedding all this logic inside the compute. This is going to be used in 3 scenarios:
- _Initialization_: Inside `createEventProcessor()`, which is a private function with called from a synchronized block in `forConsumer()` or `forDurableQueue()`. The initialization logic guarantees that there will be only one concurrent initialization, irrespective of how many threads call from outside. The initialization logic uses the output future of `getOrCreateInternalSegment`, so this is fine.
- _Regular operation_: During its operation, `tryCreateAndPinInternalSegment()` will always get a complete future related to the Event Processor at hand. This access to the map is synchronized (with the same lock as other accesses, the map itself) and it will just skip the whole logic of creating/pinning the internal Segment.
- _Shutdown_: In the worst case, shutdown can happen just after we check the event processor map. This will either be inducing an attempt to create and pin the same segment, which is not going to happen because the Segment is already created and pinned. Or the operation will fail because the Segment Container is already being shutdown.

At the moment, I don't really see an actual problem that could happen from this.

Reply:
> inside createEventProcessor(), which is a private function with called from a synchronized block in forConsumer() or forDurableQueue().

I think that statement answers my question -

Reply:
In this case putting this logic still works because 
```
CompletableFuture<EventProcessor> result) {
        return segmentSupplier.apply(name) --> this statement is executed in synchronized block
```

and so `tryCreateAndPinInternalSegment `and `eventProcessorFuture = this.eventProcessorMap.get(eventProcessorName);` happen to be executed in synchronized block.


So in this case code seems to be thread safe.